cache/pmdk_cache.cc

```diff
@@ -0,0 +1,376 @@
+    e->prev->next = e;
+    e->next->prev = e;
+    e->SetInHighPriPool(true);
+    high_pri_pool_usage_ += total_charge;
+    MaintainPoolSize();
+  } else {
+    // Insert "e" to the head of low-pri pool. Note that when
+    // high_pri_pool_ratio is 0, head of low-pri pool is also head of LRU list.
+    e->next = lru_low_pri_->next;
+    e->prev = lru_low_pri_;
+    e->prev->next = e;
+    e->next->prev = e;
+    e->SetInHighPriPool(false);
+    lru_low_pri_ = e;
+  }
+  lru_usage_ += total_charge;
+}
+
+void PMDKCacheShard::MaintainPoolSize() {
+  while (high_pri_pool_usage_ > high_pri_pool_capacity_) {
+    // Overflow last entry in high-pri pool to low-pri pool.
+    lru_low_pri_ = lru_low_pri_->next;
+    assert(lru_low_pri_ != &lru_);
+    lru_low_pri_->SetInHighPriPool(false);
+    size_t total_charge =
+        lru_low_pri_->CalcTotalCharge(metadata_charge_policy_);
+    assert(high_pri_pool_usage_ >= total_charge);
+    high_pri_pool_usage_ -= total_charge;
+  }
+}
+
+void PMDKCacheShard::EvictFromLRU(size_t charge,
+                                 autovector<TransientHandle*>* deleted) {
+  // TODO
+}
+
+void PMDKCacheShard::SetCapacity(size_t capacity) {
+  autovector<TransientHandle*> last_reference_list;
+  {
+    MutexLock l(&mutex_);
+    capacity_ = capacity;
+    high_pri_pool_capacity_ = capacity_ * high_pri_pool_ratio_;
+    EvictFromLRU(0, &last_reference_list);
+  }
+
+  // Free the entries outside of mutex for performance reasons
+  for (auto entry : last_reference_list) {
+    entry->Free();
+  }
+}
+
+void PMDKCacheShard::SetStrictCapacityLimit(bool strict_capacity_limit) {
+  MutexLock l(&mutex_);
+  strict_capacity_limit_ = strict_capacity_limit;
+}
+
+Cache::Handle* PMDKCacheShard::Lookup(const Slice& key, uint32_t hash) {
+  // TODO: lookup transient hash table
+  // lookup transient table:
+  TransientHandle* e = persistent_hashtable_->Lookup(key, hash);
+  // TODO: LRU operations.
+  return reinterpret_cast<Cache::Handle*>(e);
+}
+
+bool PMDKCacheShard::Ref(Cache::Handle* h) {
+  TransientHandle* e = reinterpret_cast<TransientHandle*>(h);
+  MutexLock l(&mutex_);
+  // To create another reference - entry must be already externally referenced
+  assert(e->HasRefs());
+  e->Ref();
+  return true;
+}
+
+void PMDKCacheShard::SetHighPriorityPoolRatio(double high_pri_pool_ratio) {
+  MutexLock l(&mutex_);
+  high_pri_pool_ratio_ = high_pri_pool_ratio;
+  high_pri_pool_capacity_ = capacity_ * high_pri_pool_ratio_;
+  MaintainPoolSize();
+}
+
+bool PMDKCacheShard::Release(Cache::Handle* handle, bool force_erase) {
+  // TODO
+  return true;
+}
+
+Status PMDKCacheShard::Insert(const Slice& key, uint32_t hash, void* value,
+                             size_t charge,
+                             void (*deleter)(const Slice& key, void* value),
+                             Cache::Handle** handle, Cache::Priority priority) {
+  Status s = Status::OK();
+  TransientHandle* e = reinterpret_cast<TransientHandle*>(
```

Comment:
`e` can be referenced by both volatile tier and persistent tier?
- If `e` is currently in volatile cache (and/or also in persistent), the persistent entry does not need to point to `e`. `e->value` points to the volatile buffer that has been used for file IO.
- `e` can be re-created when there is a hit in persistent tier, and returned to RocksDB. `e->value` points to `pentry->p_value`.

Reply:
I guess your main point is to reduce DRAM consumption by not keeping transient handles for NVM-only entries?

Reply:
I think both volatile and persistent tier referencing `e` does not consume more memory. `e` is already there anyway.
I am just trying to reason about whether we need or should let persistent tier point to `e` when it exists in volatile tier. It does not seem necessary (I may be wrong). RocksDB can always find `e` via the volatile tier.

Reply:
IIRC, if there can be a hit in persistent tier, `e` will be completely unreachable from the transient tier. Without the reference from persistent tier to `e`, each hit in persistent may result in a newly-rebuilt transient handle.

Reply:
- If `e` is still in volatile tier, then each hit will be able to find `e->value` in volatile tier. No access to persistent tier.
- If `e` is evicted from volatile tier and `free`ed, and we have a hit in persistent tier. We re-create `e`, and let `pentry->handle` point to `e`. Future hits can just use this `e`. Am I missing anything?

Reply:
Oh I thought you were saying we never have reference from `pentry` to `e`. That sounds like a good plan. In fact that may even enable us to use LRUCache as the transient tier completely. Let me think more about that.